src/network/networkd-dhcp4.c

```diff
@@ -1141,6 +1144,219 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
+
+        if (r < 0)
+                log_link_warning_errno(link, r, "Failed to update the lease lifetime: %m");
+
+        return 1;
+}
+
+static int dhcp4_load_persistent_lease(Link *link) {
+        _cleanup_(sd_dhcp_lease_unrefp) sd_dhcp_lease *lease = NULL;
+        _cleanup_free_ char *lease_file = NULL;
+        _cleanup_free_ char * lease_path = NULL;
+        int r;
+        int dir_fd;
+
+        assert(link);
+        assert(link->network);
+
+        r = link_get_dhcp_client_lease_path(link, &dir_fd, &lease_file);
+
+        /* persistent storage not avaiable/not configured, ignore */
+        if (r == -EBUSY || r == 0)
+                return r;
+
+        if (r < 0) {
+                log_link_debug_errno(link, r, "Failed to get persistent lease path: %m");
+                return r;
+        }
+
+        if (dir_fd >= 0) {
+                lease_path = path_join("/var/lib/systemd/network", lease_file);
+                if (!lease_path)
+                        return -ENOMEM;
+        }
+        else {
+                /* this shouldn't really happen, but we'll exit regardless */
+                return 0;
+        }
+
+        /* Load lease from file */
+        r = dhcp_lease_load(&lease, lease_path);
+
+        if (r == -ENOENT) {
+                /* No leases saved/exist */
+                log_link_debug(link, "No saved DHCP lease found");
+                return r;
+        }
+        if (r < 0) {
+                log_link_warning_errno(link, r, "Failed to load saved DHCP lease: %m");
+                return r;
+        }
+
+        /* check if persistent lease is expired */
+        r = dhcp4_validate_persistent_lease(link, lease);
+        if (r < 0) {
+                return r;
+        }
+
+        sd_dhcp_lease_unref(link->dhcp_lease);
+        link->dhcp_lease = TAKE_PTR(lease);
+        link_dirty(link);
+
+        log_link_info(link, "Loaded saved DHCP lease");
+
+        return 1;
+}
+
+static bool dhcp_client_persist_leases(Link *link) {
+        assert(link);
+        assert(link->manager);
+        assert(link->network);
+
+        if (link->network->dhcp_client_persist_leases >= 0)
+                return link->network->dhcp_client_persist_leases > 0;
+
+        return link->manager->dhcp_client_persist_leases > 0;
+}
+
+int link_get_dhcp_client_lease_path(Link *link, int *ret_dir_fd, char **ret_path) {
+        assert(link);
+        assert(link->network);
+        assert(ret_dir_fd);
+
+        /* This does not copy fd. Do not close fd stored in ret_dir_fd. */
+
+        if (link->network->dhcp_client_persist_leases == DHCP_CLIENT_PERSIST_LEASES_NO) {
+                *ret_dir_fd = -EBADF;
+                *ret_path = NULL;
+                return 0;
+        }
+
+        if (link->manager->persistent_storage_fd < 0)
+                return -EBUSY;
+
+        _cleanup_free_ char *p = path_join("leases", link->ifname);
```

Comment:
I'd like to name this as "dhcp-client-leases" or so, to make it consistent with dhcp-server-leases, and also for DHCPv6 leases in the future.